File: vscode-mmemu-lua/src/test-persistence.ts
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs/promises';
import { execSync } from 'child_process';

/**
 * Single test run result
 */
export interface TestRunResult {
    timestamp: string;    // ISO 8601
    gitHash: string;      // Commit hash (7 chars)
    status: 'pass' | 'fail' | 'skip';
    duration: number;     // Milliseconds
    error?: string;       // Error message if failed
}

/**
 * Test history entry
 */
export interface TestHistory {
    file: string;
    history: TestRunResult[];
    lastStatus: 'pass' | 'fail' | 'skip';
    lastGitHash: string;
    statusChanges: number;
}

/**
 * Persistence data structure
 */
interface TestPersistenceData {
    version: string;
    workspace: string;
    tests: Record<string, TestHistory>;
}

/**
 * Test Run Persistence Manager
 * Tracks test results with git hash recording
 */
export class TestPersistence {
    private dataPath: string;
    private data: TestPersistenceData;
    private dirty = false;

    constructor(workspaceRoot?: string) {
        this.dataPath = this.getPersistencePath();
        this.data = {
            version: '1.0',
            workspace: workspaceRoot || vscode.workspace.workspaceFolders?.[0]?.uri.fsPath || '',
            tests: {},
        };
        this.loadData();
    }

    /**
     * Get persistence file path
     */
    private getPersistencePath(): string {
        let basePath: string;

        if (process.platform === 'win32') {
            basePath = process.env.APPDATA || path.join(process.env.USERPROFILE || '', 'AppData', 'Roaming');
        } else if (process.platform === 'darwin') {
            basePath = path.join(process.env.HOME || '', 'Library', 'Application Support');
        } else {
            basePath = path.join(process.env.HOME || '', '.local', 'share');
        }

        const mmemuDir = path.join(basePath, 'mmemu');
        return path.join(mmemuDir, 'test-results.json');
    }

    /**
     * Load test results from file
     */
    private async loadData(): Promise<void> {
        try {
            const content = await fs.readFile(this.dataPath, 'utf-8');
            const loaded = JSON.parse(content);

            // Validate structure
            if (loaded.version && loaded.tests) {
                this.data = loaded;
                this.cleanupOldData();
            }
        } catch (error) {
            // File doesn't exist or is invalid - start fresh
            this.data = {
                version: '1.0',
                workspace: vscode.workspace.workspaceFolders?.[0]?.uri.fsPath || '',
                tests: {},
            };
        }
    }

    /**
     * Save test results to file
     */
    private async saveData(): Promise<void> {
        try {
            const dir = path.dirname(this.dataPath);
            await fs.mkdir(dir, { recursive: true });
            await fs.writeFile(this.dataPath, JSON.stringify(this.data, null, 2));
            this.dirty = false;
        } catch (error) {
            console.error('Failed to save test persistence data:', error);
        }
    }

    /**
     * Record a test run result
     * Returns true if test status changed
     */
    async recordTestRun(testName: string, file: string, result: Omit<TestRunResult, 'gitHash'>): Promise<boolean> {
        // Get current git hash
        const gitHash = this.getGitHash();

        // Add git hash to result
        const fullResult: TestRunResult = {
            ...result,
            gitHash,
        };

        // Check if status changed
        const hasChanged = this.hasStatusChanged(testName, result.status);

        // Initialize test history if needed
        if (!this.data.tests[testName]) {
            this.data.tests[testName] = {
                file,
                history: [],
                lastStatus: result.status,
                lastGitHash: gitHash,
                statusChanges: 0,
            };
        }

        const history = this.data.tests[testName];

        // Record the run
        history.history.push(fullResult);

        // Update status and count changes
        if (hasChanged) {
            history.statusChanges++;
        }

        history.lastStatus = result.status;
        history.lastGitHash = gitHash;

        this.dirty = true;
        await this.saveData();

        return hasChanged;
    }

    /**
     * Check if test status has changed
     */
    private hasStatusChanged(testName: string, newStatus: string): boolean {
        if (!this.data.tests[testName]) return true;
        return this.data.tests[testName].lastStatus !== newStatus;
    }

    /**
     * Get test history
     */
    getTestHistory(testName: string): TestRunResult[] {
        return this.data.tests[testName]?.history || [];
    }

    /**
     * Get test status change count
     */
    getStatusChangeCount(testName: string): number {
        return this.data.tests[testName]?.statusChanges || 0;
    }

    /**
     * Get last git hash for test
     */
    getLastGitHash(testName: string): string | undefined {
        return this.data.tests[testName]?.lastGitHash;
    }

    /**
     * Get last status for test
     */
    getLastStatus(testName: string): 'pass' | 'fail' | 'skip' | undefined {
        return this.data.tests[testName]?.lastStatus;
    }

    /**
     * Get current git commit hash
     */
    private getGitHash(): string {
        try {
            const hash = execSync('git rev-parse --short HEAD', {
                cwd: this.data.workspace,
                encoding: 'utf-8',
                stdio: ['pipe', 'pipe', 'pipe'],
            }).trim();
            return hash;
        } catch {
            return 'unknown';
        }
    }

    /**
     * Clean up old data (>30 days old)
     */
    private cleanupOldData(): void {
        const thirtyDaysAgo = new Date();
        thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);

        for (const testName in this.data.tests) {
            const history = this.data.tests[testName];

            // Keep at least 10 most recent entries
            if (history.history.length > 10) {
                const filtered = history.history.filter(
                    run => new Date(run.timestamp) > thirtyDaysAgo
                );

                // If we'd delete all old data, keep last 10
                if (filtered.length < 10) {
                    history.history = history.history.slice(-10);
                } else {
                    history.history = filtered;
                }

                this.dirty = true;
            }
        }

        if (this.dirty) {
            this.saveData();
        }
    }

    /**
     * Clear all test history
     */
    async clearAllHistory(): Promise<void> {
        this.data.tests = {};
        this.dirty = true;
        await this.saveData();
    }

    /**
     * Clear history for specific test
     */
    async clearTestHistory(testName: string): Promise<void> {
        delete this.data.tests[testName];
        this.dirty = true;
        await this.saveData();
    }

    /**
     * Get all test names
     */
    getTestNames(): string[] {
        return Object.keys(this.data.tests);
    }

    /**
     * Export history as CSV
     */
    exportAsCSV(testName: string): string {
        const history = this.getTestHistory(testName);

        const rows = [
            'timestamp,git_hash,status,duration_ms,error',
            ...history.map(run =>
                `${run.timestamp},${run.gitHash},${run.status},${run.duration},"${run.error || ''}"`
            ),
        ];

        return rows.join('\n');
    }

    /**
     * Export all history as CSV
     */
    exportAllAsCSV(): string {
        const rows = [
            'test_name,timestamp,git_hash,status,duration_ms,error',
        ];

        for (const testName in this.data.tests) {
            const history = this.data.tests[testName];
            for (const run of history.history) {
                rows.push(
                    `${testName},${run.timestamp},${run.gitHash},${run.status},${run.duration},"${run.error || ''}"`
                );
            }
        }

        return rows.join('\n');
    }

    /**
     * Get test status description
     */
    getStatusDescription(testName: string): string {
        const test = this.data.tests[testName];
        if (!test) return 'No history';

        const lastRun = test.history[test.history.length - 1];
        if (!lastRun) return 'No history';

        const age = Math.round((Date.now() - new Date(lastRun.timestamp).getTime()) / 1000);
        const ageStr = age < 60 ? `${age}s ago` : `${Math.round(age / 60)}m ago`;

        return `${test.lastStatus === 'pass' ? '✓' : '✗'} ${test.lastStatus} (${ageStr}, ${test.lastGitHash})`;
    }
}

File: src/libdebug/main/test_persistence.cpp
#include "test_persistence.h"
#include <filesystem>
#include <fstream>
#include <cstdlib>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <cstring>
#include <array>
#include <chrono>

namespace fs = std::filesystem;

/**
 * TestRunResult JSON conversion
 */
json TestRunResult::toJson() const {
    return json{
        {"timestamp", timestamp},
        {"gitHash", gitHash},
        {"status", status},
        {"duration", duration},
        {"error", error}
    };
}

TestRunResult TestRunResult::fromJson(const json& j) {
    TestRunResult result;
    result.timestamp = j.value("timestamp", "");
    result.gitHash = j.value("gitHash", "unknown");
    result.status = j.value("status", "");
    result.duration = j.value("duration", 0u);
    result.error = j.value("error", "");
    return result;
}

/**
 * TestHistory JSON conversion
 */
json TestHistory::toJson() const {
    json historyArray = json::array();
    for (const auto& run : history) {
        historyArray.push_back(run.toJson());
    }

    return json{
        {"file", file},
        {"history", historyArray},
        {"lastStatus", lastStatus},
        {"lastGitHash", lastGitHash},
        {"statusChanges", statusChanges}
    };
}

TestHistory TestHistory::fromJson(const json& j) {
    TestHistory history;
    history.file = j.value("file", "");
    history.lastStatus = j.value("lastStatus", "");
    history.lastGitHash = j.value("lastGitHash", "unknown");
    history.statusChanges = j.value("statusChanges", 0u);

    if (j.contains("history") && j["history"].is_array()) {
        for (const auto& run : j["history"]) {
            history.history.push_back(TestRunResult::fromJson(run));
        }
    }

    return history;
}

/**
 * Constructor
 */
TestPersistence::TestPersistence(const std::string& workspaceRoot)
    : workspaceRoot(workspaceRoot.empty() ? "." : workspaceRoot) {
    dataPath = getPersistencePath();
    loadData();
}

/**
 * Get persistence file path (platform-aware)
 */
std::string TestPersistence::getPersistencePath() {
    std::string basePath;

#ifdef _WIN32
    const char* appData = std::getenv("APPDATA");
    basePath = appData ? appData : ".";
#elif __APPLE__
    const char* home = std::getenv("HOME");
    basePath = std::string(home ? home : ".") + "/Library/Application Support";
#else
    const char* home = std::getenv("HOME");
    basePath = std::string(home ? home : ".") + "/.local/share";
#endif

    fs::path mmemuPath = fs::path(basePath) / "mmemu";
    fs::create_directories(mmemuPath);

    return (mmemuPath / "test-results.json").string();
}

/**
 * Load test results from file
 */
void TestPersistence::loadData() {
    try {
        if (!fs::exists(dataPath)) {
            return;
        }

        std::ifstream file(dataPath);
        if (!file.is_open()) {
            return;
        }

        json data;
        file >> data;

        if (data.contains("tests") && data["tests"].is_object()) {
            for (auto& [testName, testData] : data["tests"].items()) {
                tests[testName] = TestHistory::fromJson(testData);
            }
        }

        cleanupOldData();
    } catch (const std::exception& e) {
        // Failed to load - start fresh
        tests.clear();
    }
}

/**
 * Save test results to file
 */
void TestPersistence::saveData() {
    try {
        json data;
        data["version"] = "1.0";
        data["workspace"] = workspaceRoot;
        data["tests"] = json::object();

        for (const auto& [testName, history] : tests) {
            data["tests"][testName] = history.toJson();
        }

        fs::create_directories(fs::path(dataPath).parent_path());

        std::ofstream file(dataPath);
        file << data.dump(2) << std::endl;
        dirty = false;
    } catch (const std::exception& e) {
        // Failed to save - silently continue
    }
}

/**
 * Record test run result
 */
bool TestPersistence::recordTestRun(const std::string& testName,
                                     const std::string& filePath,
                                     const TestRunResult& result) {
    TestRunResult fullResult = result;
    if (fullResult.gitHash.empty()) {
        fullResult.gitHash = getGitHash();
    }
    if (fullResult.timestamp.empty()) {
        fullResult.timestamp = getCurrentTimestamp();
    }

    bool hasChanged = hasStatusChanged(testName, result.status);

    if (tests.find(testName) == tests.end()) {
        TestHistory history;
        history.file = filePath;
        history.lastStatus = result.status;
        history.lastGitHash = fullResult.gitHash;
        history.statusChanges = 0;
        tests[testName] = history;
    }

    auto& history = tests[testName];
    history.history.push_back(fullResult);

    // Only increment status changes if there was a prior status and it changed
    if (hasChanged && history.history.size() > 1) {
        history.statusChanges++;
    }

    history.lastStatus = result.status;
    history.lastGitHash = fullResult.gitHash;

    dirty = true;
    saveData();

    return hasChanged;
}

/**
 * Check if status has changed
 */
bool TestPersistence::hasStatusChanged(const std::string& testName,
                                        const std::string& newStatus) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return true;
    return it->second.lastStatus != newStatus;
}

/**
 * Get test history
 */
std::vector<TestRunResult> TestPersistence::getTestHistory(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return {};
    return it->second.history;
}

/**
 * Get status change count
 */
uint32_t TestPersistence::getStatusChangeCount(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return 0;
    return it->second.statusChanges;
}

/**
 * Get last status
 */
std::string TestPersistence::getLastStatus(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return "";
    return it->second.lastStatus;
}

/**
 * Get last git hash
 */
std::string TestPersistence::getLastGitHash(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return "unknown";
    return it->second.lastGitHash;
}

/**
 * Get status description
 */
std::string TestPersistence::getStatusDescription(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return "No history";

    const auto& history = it->second;
    if (history.history.empty()) return "No history";

    const auto& lastRun = history.history.back();

    // Calculate age
    auto now = std::chrono::system_clock::now();
    auto lastTime = std::chrono::system_clock::from_time_t(
        std::chrono::system_clock::to_time_t(now)  // Simplified; actual parsing from ISO would be better
    );

    std::string statusSymbol = history.lastStatus == "pass" ? "✓" : "✗";
    std::string result = statusSymbol + " " + history.lastStatus + " (" +
                         history.lastGitHash + ")";

    return result;
}

/**
 * Export as CSV (single test)
 */
std::string TestPersistence::exportAsCSV(const std::string& testName) const {
    auto it = tests.find(testName);
    if (it == tests.end()) return "timestamp,git_hash,status,duration_ms,error\n";

    std::string csv = "timestamp,git_hash,status,duration_ms,error\n";

    for (const auto& run : it->second.history) {
        csv += run.timestamp + "," + run.gitHash + "," + run.status + "," +
               std::to_string(run.duration) + ",\"" + run.error + "\"\n";
    }

    return csv;
}

/**
 * Export all as CSV
 */
std::string TestPersistence::exportAllAsCSV() const {
    std::string csv = "test_name,timestamp,git_hash,status,duration_ms,error\n";

    for (const auto& [testName, history] : tests) {
        for (const auto& run : history.history) {
            csv += testName + "," + run.timestamp + "," + run.gitHash + "," +
                   run.status + "," + std::to_string(run.duration) + ",\"" +
                   run.error + "\"\n";
        }
    }

    return csv;
}

/**
 * Clear all history
 */
void TestPersistence::clearAllHistory() {
    tests.clear();
    dirty = true;
    saveData();
}

/**
 * Clear test history
 */
void TestPersistence::clearTestHistory(const std::string& testName) {
    tests.erase(testName);
    dirty = true;
    saveData();
}

/**
 * Get test names
 */
std::vector<std::string> TestPersistence::getTestNames() const {
    std::vector<std::string> names;
    for (const auto& [name, _] : tests) {
        names.push_back(name);
    }
    return names;
}

/**
 * Get current git hash
 */
std::string TestPersistence::getGitHash() {
    std::array<char, 128> buffer;
    std::string result;

    FILE* pipe = popen("git rev-parse --short HEAD 2>/dev/null", "r");
    if (!pipe) return "unknown";

    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
        result += buffer.data();
    }

    pclose(pipe);

    // Trim whitespace
    if (!result.empty() && result.back() == '\n') {
        result.pop_back();
    }

    return result.empty() ? "unknown" : result;
}

/**
 * Get current timestamp (ISO 8601)
 */
std::string TestPersistence::getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);

    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time), "%Y-%m-%dT%H:%M:%SZ");
    return ss.str();
}

/**
 * Clean up old data
 */
void TestPersistence::cleanupOldData() {
    auto now = std::chrono::system_clock::now();
    auto thirtyDaysAgo = now - std::chrono::hours(30 * 24);
    auto thirtyDaysAgoTime = std::chrono::system_clock::to_time_t(thirtyDaysAgo);

    for (auto& [testName, history] : tests) {
        if (history.history.size() > 10) {
            std::vector<TestRunResult> filtered;

            for (const auto& run : history.history) {
                // Parse ISO timestamp (simplified - assumes valid format)
                // Full implementation would parse the timestamp properly
                filtered.push_back(run);
            }

            // Keep last 10 if too much was filtered
            if (filtered.size() < 10 && history.history.size() > 10) {
                history.history.erase(
                    history.history.begin(),
                    history.history.end() - 10
                );
            } else if (filtered.size() >= 10) {
                history.history = filtered;
            }

            dirty = true;
        }
    }

    if (dirty) {
        saveData();
    }
}
